ksql-engine/src/main/java/io/confluent/ksql/function/udf/geo/GeoDistanceKudf.java

```diff
@@ -0,0 +1,100 @@
+ * formula for "great circle distance". The 2 input points should be specified as (lat, lon) pairs,
+ * measured in decimal degrees.
+ * 
+ * <p>An optional fifth parameter allows to specify either "MI" (miles) or "KM" (kilometers) as the
+ * desired unit for the output measurement. Default is KM.
+ *
+ */
+public class GeoDistanceKudf implements Kudf {
+
+  // effective value of Earth radius (note we technically live on a slightly squashed sphere, not
+  // a truly round one, so different authorities will quote slightly different values for the 'best'
+  // value to use as effective radius. The difference between the 2 most commonly-quoted values
+  // measures out to about 0.1% in most real-world cases, which is within the margin of error of
+  // using this kind of great-circle methodology anyway (~0.5%).
+  private static final double EARTH_RADIUS_KM = 6371;
+  private static final double EARTH_RADIUS_MILES = 3959;
+
+
+  @Override
+  public Object evaluate(Object... args) {
+    if ((args.length < 4) || (args.length > 5)) {
+      throw new KsqlFunctionException(
+          "GeoDistance function expects either 4 or 5 arguments: lat1, lon1, lat2, lon2, (MI/KM)");
+    }
+
+    double lat1 = ((Number) args[0]).doubleValue();
+    double lon1 = ((Number) args[1]).doubleValue();
+    double lat2 = ((Number) args[2]).doubleValue();
+    double lon2 = ((Number) args[3]).doubleValue();
+    checkParameterValuesAreLegal(lat1, lon1, lat2, lon2);
+    double chosenRadius = selectEarthRadiusToUse(args);
+
+    double deltaLat = Math.toRadians(lat2 - lat1);
+    double deltaLon = Math.toRadians(lon2 - lon1);
+
+    double lat1Radians = Math.toRadians(lat1);
+    double lat2Radians = Math.toRadians(lat2);
+
+    double a =
+        haversin(deltaLat) + haversin(deltaLon) * Math.cos(lat1Radians) * Math.cos(lat2Radians);
+    double distanceInRadians = 2 * Math.asin(Math.sqrt(a));
+    return distanceInRadians * chosenRadius;
+  }
+
+  private void checkParameterValuesAreLegal(double lat1, double lon1, double lat2, double lon2) {
+    if (lat1 < 0 || lat2 < 0 || lat1 > 90 || lat2 > 90) {
+      throw new KsqlFunctionException(
+          "valid latitude values for GeoDistance function are in the range of 0 to 90"
+              + " decimal degrees");
+    }
+    if (lon1 < -180 || lon2 < -180 || lon1 > 180 || lon2 > 180) {
+      throw new KsqlFunctionException(
+          "valid longitude values for GeoDistance function are in the range of -180 to +180"
+              + " decimal degrees");
+    }
+  }
+
+  private double selectEarthRadiusToUse(Object... args) {
+    double chosenRadius = EARTH_RADIUS_KM;
+    if (args.length == 5) {
+      String outputUnit = args[4].toString().toLowerCase();
+      if (outputUnit.equals("mi") || outputUnit.startsWith("mile")) {
+        chosenRadius = EARTH_RADIUS_MILES;
+      } else if (outputUnit.equals("km") || outputUnit.startsWith("kilomet")) {
```

Comment:
nit: `equals("kilometeres")` ?
similarly above for `mile`. Only as otherwise the valid values for this could `kilometasdfhaklsjfhak` which makes no sense!

Reply:
so my thinking here was 2-fold: first up, i wanted to avoid having to force people to remember whether it's singular or plural distance units (similar to how we do for time units, for example); secondarily, to accommodate folks in places where they like to write "kilometre" and not force them to use the anglicized spelling all the time. Would you guys prefer this be changed to test for all the allowable permutations explicitly ?